1.train_models/log_reg_plates_cp_fs_data/plate_4/nbconverted/plate_4_classify_genotype.py

```diff
@@ -0,0 +1,248 @@
+from sklearn.exceptions import ConvergenceWarning
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import RandomizedSearchCV, train_test_split
+from sklearn.preprocessing import LabelEncoder
+from sklearn.utils import parallel_backend
+
+# ## Find the root of the git repo on the host system
+
+# In[2]:
+
+
+# Get the current working directory
+cwd = pathlib.Path.cwd()
+
+if (cwd / ".git").is_dir():
+    root_dir = cwd
+
+else:
+    root_dir = None
+    for parent in cwd.parents:
+        if (parent / ".git").is_dir():
+            root_dir = parent
+            break
+
+# Check if a Git root directory was found
+if root_dir is None:
+    raise FileNotFoundError("No Git root directory found.")
+
+
+# ## Define and create paths
+
+# In[3]:
+
+
+# Input paths
+plate_path = "Plate_4_sc_norm_fs.parquet"
+plate_path = pathlib.Path(f"{root_dir}/nf1_painting_repo/3.processing_features/data/feature_selected_data/{plate_path}")
+
+# Output paths
+models_path = pathlib.Path("trained_models")
+data_path = pathlib.Path("model_data")
+
+# Create output paths if nonexistent
+models_path.mkdir(parents=True, exist_ok=True)
+data_path.mkdir(parents=True, exist_ok=True)
+
+
+# ## Load the dataframe
+
+# In[4]:
+
+
+platedf = pd.read_parquet(plate_path)
+
+
+# ## Define column names
+
+# In[5]:
+
+
+well_column = "Metadata_Well"
+gene_column = "Metadata_genotype"
+concentration_column = "Metadata_Concentration"
+
+
+# # Process data
+
+# ## Down-sample by genotype
+
+# In[6]:
+
+
+# Filter data where the siRNA construct concentration is zero
+platedf = platedf[platedf[concentration_column] == 0]
+
+min_gene = platedf[gene_column].value_counts().min()
+platedf = platedf.groupby(gene_column, group_keys=False).apply(lambda x: x.sample(n=min_gene, random_state=0))
+
+
+# ## Encode genotypes
+
+# In[7]:
+
+
+# Encode classes
+le = LabelEncoder()
+featdf = platedf.copy()
+featdf[gene_column] = le.fit_transform(featdf[gene_column])
+
+
+# ## Remove Metadata and encode labels
+
+# In[8]:
+
+
+# Create dataframe without Metadata
+feat_columns = [col for col in platedf.columns if "Metadata" not in col]
+featdf = platedf[feat_columns]
+
+# Encode labels
+platedf[gene_column] = le.transform(platedf[gene_column])
+
+
+# ## Split the Data
+
+# In[9]:
+
+
+test_frac = 0.2
+val_frac = 0.15
+
+X_train, X_test, y_train, y_test = train_test_split(featdf, platedf[gene_column], test_size = test_frac, random_state=0, shuffle=True, stratify=platedf[gene_column])
+
+
+# ## Shuffle the training data for the shuffled model
+
+# In[10]:
+
+
+X_train_shuf = X_train.copy()
+
+rng = np.random.default_rng(0)
+
+for column in X_train_shuf.columns:
+    X_train_shuf[column] = rng.permutation(X_train_shuf[column])
+
+
+# # Train Models
+
+# ## Define model variables
+
+# In[11]:
+
+
+# Define the hyperparameter search space
+param_dist = {
+    'C': uniform(loc=0.5, scale=1.5),
+    "l1_ratio": uniform(loc=0, scale=1),
+}
+
+# Create a Logistic Regression model
```

Comment:
consider adding inline citation for Roshan's phenotypic model if this is where you got this code from.
"Code adapted from:...."

Reply:
I don't remember if it was exactly the code, or just the same approach, but good to think about.